File: reference/go/ws.go
// Package caretv4 is the caret/v4 reference backend: the transport core,
// the shared live-audio lifecycle, the pluggable lanes, and the
// black-box conformance checker.
//
// This file is the WebSocket layer. RFC 6455 is small enough that
// hand-writing the useful half of it costs less than a dependency: a
// handshake, a frame header, a mask, and the four opcodes the protocol
// uses. Nothing here is general-purpose — it does exactly what caret/v4
// needs (text control frames, binary audio, ping/pong keep-alive, a
// close code) and nothing else. Extensions and continuation-heavy
// senders are handled, compression is not negotiated.
package caretv4

import (
	"bufio"
	"crypto/rand"
	"crypto/sha1"
	"crypto/tls"
	"encoding/base64"
	"encoding/binary"
	"errors"
	"fmt"
	"io"
	"net"
	"net/http"
	"net/url"
	"strings"
	"sync"
	"time"
)

// wsGUID is the RFC 6455 handshake constant.
// closeDrain bounds how long Close spends reading the peer out before
// closing the socket. Skipped once the peer's own close frame has
// arrived, since there is then nothing left in flight.
const closeDrain = 500 * time.Millisecond

const wsGUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

// WebSocket opcodes.
const (
	opContinuation = 0x0
	opText         = 0x1
	opBinary       = 0x2
	opClose        = 0x8
	opPing         = 0x9
	opPong         = 0xA
)

// maxPayload caps a single inbound message. The protocol's own frame
// bound is smaller (524 288 bytes by default); this is the transport
// backstop that keeps a hostile length header from allocating a gigabyte.
const maxPayload = 8 << 20

// ErrClosed is returned by ReadMessage when the peer sent a close frame.
// The close code, if any, is carried by CloseError.
var ErrClosed = errors.New("websocket: peer closed")

// CloseError reports the code and reason the peer closed with.
type CloseError struct {
	Code   uint16
	Reason string
}

func (e *CloseError) Error() string {
	return fmt.Sprintf("websocket: closed %d %q", e.Code, e.Reason)
}

func (e *CloseError) Unwrap() error { return ErrClosed }

// Conn is one WebSocket connection. Reads happen on one goroutine;
// writes are serialized by a mutex so a progress ticker and a partial
// emitter can share the socket.
type Conn struct {
	conn     net.Conn
	br       *bufio.Reader
	isClient bool

	wmu      sync.Mutex
	closed   bool
	sawClose bool
	closeMu  sync.Mutex
}

// Message is one complete inbound WebSocket message.
type Message struct {
	Binary bool
	Data   []byte
}

func acceptKey(key string) string {
	sum := sha1.Sum([]byte(key + wsGUID))
	return base64.StdEncoding.EncodeToString(sum[:])
}

// Upgrade completes the server side of the handshake and hijacks the
// connection. On failure it has already written an HTTP error response.
func Upgrade(w http.ResponseWriter, r *http.Request) (*Conn, error) {
	if !strings.EqualFold(r.Header.Get("Upgrade"), "websocket") ||
		!headerContainsToken(r.Header.Get("Connection"), "upgrade") {
		http.Error(w, "expected a websocket upgrade", http.StatusBadRequest)
		return nil, errors.New("websocket: not an upgrade request")
	}
	if r.Header.Get("Sec-WebSocket-Version") != "13" {
		w.Header().Set("Sec-WebSocket-Version", "13")
		http.Error(w, "unsupported websocket version", http.StatusUpgradeRequired)
		return nil, errors.New("websocket: bad version")
	}
	key := r.Header.Get("Sec-WebSocket-Key")
	if key == "" {
		http.Error(w, "missing Sec-WebSocket-Key", http.StatusBadRequest)
		return nil, errors.New("websocket: missing key")
	}
	hj, ok := w.(http.Hijacker)
	if !ok {
		http.Error(w, "connection cannot be hijacked", http.StatusInternalServerError)
		return nil, errors.New("websocket: response writer is not a Hijacker")
	}
	netConn, rw, err := hj.Hijack()
	if err != nil {
		return nil, fmt.Errorf("websocket: hijack: %w", err)
	}
	_ = netConn.SetDeadline(time.Time{})
	resp := "HTTP/1.1 101 Switching Protocols\r\n" +
		"Upgrade: websocket\r\n" +
		"Connection: Upgrade\r\n" +
		"Sec-WebSocket-Accept: " + acceptKey(key) + "\r\n\r\n"
	if _, err := rw.WriteString(resp); err != nil {
		_ = netConn.Close()
		return nil, fmt.Errorf("websocket: write handshake: %w", err)
	}
	if err := rw.Flush(); err != nil {
		_ = netConn.Close()
		return nil, fmt.Errorf("websocket: flush handshake: %w", err)
	}
	return &Conn{conn: netConn, br: rw.Reader}, nil
}

func headerContainsToken(header, token string) bool {
	for _, part := range strings.Split(header, ",") {
		if strings.EqualFold(strings.TrimSpace(part), token) {
			return true
		}
	}
	return false
}

// DialOptions configures the client handshake.
type DialOptions struct {
	Header    http.Header
	TLSConfig *tls.Config
	Timeout   time.Duration
}

// Dial opens a client connection to a ws:// or wss:// URL. It returns
// the handshake response so a caller can inspect a non-101 status —
// caret/v4 backends are allowed to refuse the upgrade with HTTP 401.
func Dial(rawURL string, opts DialOptions) (*Conn, *http.Response, error) {
	u, err := url.Parse(rawURL)
	if err != nil {
		return nil, nil, fmt.Errorf("websocket: bad url: %w", err)
	}
	secure := false
	switch u.Scheme {
	case "ws":
	case "wss":
		secure = true
	default:
		return nil, nil, fmt.Errorf("websocket: unsupported scheme %q", u.Scheme)
	}
	host := u.Host
	if u.Port() == "" {
		if secure {
			host = net.JoinHostPort(u.Hostname(), "443")
		} else {
			host = net.JoinHostPort(u.Hostname(), "80")
		}
	}
	timeout := opts.Timeout
	if timeout <= 0 {
		timeout = 15 * time.Second
	}
	dialer := &net.Dialer{Timeout: timeout}
	var netConn net.Conn
	if secure {
		cfg := opts.TLSConfig
		if cfg == nil {
			cfg = &tls.Config{}
		}
		cfg = cfg.Clone()
		if cfg.ServerName == "" {
			cfg.ServerName = u.Hostname()
		}
		netConn, err = tls.DialWithDialer(dialer, "tcp", host, cfg)
	} else {
		netConn, err = dialer.Dial("tcp", host)
	}
	if err != nil {
		return nil, nil, fmt.Errorf("websocket: dial: %w", err)
	}

	var raw [16]byte
	if _, err := rand.Read(raw[:]); err != nil {
		_ = netConn.Close()
		return nil, nil, err
	}
	key := base64.StdEncoding.EncodeToString(raw[:])

	path := u.RequestURI()
	var req strings.Builder
	fmt.Fprintf(&req, "GET %s HTTP/1.1\r\n", path)
	fmt.Fprintf(&req, "Host: %s\r\n", u.Host)
	req.WriteString("Upgrade: websocket\r\nConnection: Upgrade\r\n")
	fmt.Fprintf(&req, "Sec-WebSocket-Key: %s\r\n", key)
	req.WriteString("Sec-WebSocket-Version: 13\r\n")
	for name, values := range opts.Header {
		for _, v := range values {
			fmt.Fprintf(&req, "%s: %s\r\n", name, v)
		}
	}
	req.WriteString("\r\n")

	_ = netConn.SetDeadline(time.Now().Add(timeout))
	if _, err := io.WriteString(netConn, req.String()); err != nil {
		_ = netConn.Close()
		return nil, nil, fmt.Errorf("websocket: write handshake: %w", err)
	}
	br := bufio.NewReader(netConn)
	httpReq, _ := http.NewRequest(http.MethodGet, rawURL, nil)
	resp, err := http.ReadResponse(br, httpReq)
	if err != nil {
		_ = netConn.Close()
		return nil, nil, fmt.Errorf("websocket: read handshake: %w", err)
	}
	if resp.StatusCode != http.StatusSwitchingProtocols {
		body, _ := io.ReadAll(io.LimitReader(resp.Body, 4096))
		resp.Body = io.NopCloser(strings.NewReader(string(body)))
		_ = netConn.Close()
		return nil, resp, fmt.Errorf("websocket: handshake refused with HTTP %d", resp.StatusCode)
	}
	if resp.Header.Get("Sec-WebSocket-Accept") != acceptKey(key) {
		_ = netConn.Close()
		return nil, resp, errors.New("websocket: bad Sec-WebSocket-Accept")
	}
	_ = netConn.SetDeadline(time.Time{})
	return &Conn{conn: netConn, br: br, isClient: true}, resp, nil
}

// SetReadDeadline bounds the next ReadMessage. A zero time clears it.
func (c *Conn) SetReadDeadline(t time.Time) error { return c.conn.SetReadDeadline(t) }

// RemoteAddr reports the peer address, for logging ids only.
func (c *Conn) RemoteAddr() net.Addr { return c.conn.RemoteAddr() }

// ReadMessage returns the next complete text or binary message,
// answering pings and skipping pongs on the way. A close frame from the
// peer surfaces as a *CloseError wrapping ErrClosed.
func (c *Conn) ReadMessage() (Message, error) {
	var assembled []byte
	var msgBinary bool
	inFragment := false
	for {
		fin, opcode, payload, err := c.readFrame()
		if err != nil {
			return Message{}, err
		}
		switch opcode {
		case opPing:
			_ = c.writeFrame(opPong, payload)
			continue
		case opPong:
			continue
		case opClose:
			code := uint16(1005)
			reason := ""
			if len(payload) >= 2 {
				code = binary.BigEndian.Uint16(payload[:2])
				reason = string(payload[2:])
			}
			c.closeMu.Lock()
			c.sawClose = true
			c.closeMu.Unlock()
			return Message{}, &CloseError{Code: code, Reason: reason}
		case opText, opBinary:
			if inFragment {
				return Message{}, errors.New("websocket: interleaved data frame")
			}
			msgBinary = opcode == opBinary
			assembled = payload
		case opContinuation:
			if !inFragment {
				return Message{}, errors.New("websocket: continuation without start")
			}
			assembled = append(assembled, payload...)
		default:
			return Message{}, fmt.Errorf("websocket: unknown opcode %d", opcode)
		}
		if fin {
			return Message{Binary: msgBinary, Data: assembled}, nil
		}
		inFragment = true
	}
}

func (c *Conn) readFrame() (fin bool, opcode byte, payload []byte, err error) {
	var head [2]byte
	if _, err = io.ReadFull(c.br, head[:]); err != nil {
		return
	}
	fin = head[0]&0x80 != 0
	opcode = head[0] & 0x0F
	masked := head[1]&0x80 != 0
	length := uint64(head[1] & 0x7F)
	switch length {
	case 126:
		var ext [2]byte
		if _, err = io.ReadFull(c.br, ext[:]); err != nil {
			return
		}
		length = uint64(binary.BigEndian.Uint16(ext[:]))
	case 127:
		var ext [8]byte
		if _, err = io.ReadFull(c.br, ext[:]); err != nil {
			return
		}
		length = binary.BigEndian.Uint64(ext[:])
	}
	if length > maxPayload {
		err = fmt.Errorf("websocket: frame of %d bytes exceeds the transport cap", length)
		return
	}
	if !c.isClient && !masked {
		err = errors.New("websocket: client frame was not masked")
		return
	}
	var mask [4]byte
	if masked {
		if _, err = io.ReadFull(c.br, mask[:]); err != nil {
			return
		}
	}
	payload = make([]byte, length)
	if _, err = io.ReadFull(c.br, payload); err != nil {
		return
	}
	if masked {
		for i := range payload {
			payload[i] ^= mask[i%4]
		}
	}
	return
}

func (c *Conn) writeFrame(opcode byte, payload []byte) error {
	c.wmu.Lock()
	defer c.wmu.Unlock()
	header := make([]byte, 0, 14)
	header = append(header, 0x80|opcode)
	maskBit := byte(0)
	if c.isClient {
		maskBit = 0x80
	}
	switch n := len(payload); {
	case n < 126:
		header = append(header, maskBit|byte(n))
	case n <= 0xFFFF:
		header = append(header, maskBit|126, byte(n>>8), byte(n))
	default:
		header = append(header, maskBit|127)
		var ext [8]byte
		binary.BigEndian.PutUint64(ext[:], uint64(n))
		header = append(header, ext[:]...)
	}
	body := payload
	if c.isClient {
		var mask [4]byte
		if _, err := rand.Read(mask[:]); err != nil {
			return err
		}
		header = append(header, mask[:]...)
		body = make([]byte, len(payload))
		for i := range payload {
			body[i] = payload[i] ^ mask[i%4]
		}
	}
	if _, err := c.conn.Write(append(header, body...)); err != nil {
		return err
	}
	return nil
}

// WriteText sends one text frame.
func (c *Conn) WriteText(s string) error { return c.writeFrame(opText, []byte(s)) }

// WriteBinary sends one binary frame.
func (c *Conn) WriteBinary(b []byte) error { return c.writeFrame(opBinary, b) }

// WritePing sends an empty ping, the cheap keep-alive the protocol
// accepts in place of a progress event.
func (c *Conn) WritePing() error { return c.writeFrame(opPing, nil) }

// Close sends a close frame with the given code and shuts the socket
// down. Calling it twice is harmless.
func (c *Conn) Close(code uint16, reason string) error {
	c.closeMu.Lock()
	if c.closed {
		c.closeMu.Unlock()
		return nil
	}
	c.closed = true
	drain := !c.sawClose
	c.closeMu.Unlock()

	payload := make([]byte, 2, 2+len(reason))
	binary.BigEndian.PutUint16(payload, code)
	payload = append(payload, reason...)
	_ = c.writeFrame(opClose, payload)
	// Drain whatever the peer is still sending before closing the
	// socket. This is not politeness: a TCP socket closed with unread
	// inbound data is reset, and a reset can discard the peer's receive
	// buffer along with the terminal event we just wrote. §8 lets a
	// backend answer a replayed operation from cache the moment it says
	// ready, while the client is mid-upload, so this path is ordinary
	// rather than exotic.
	if drain {
		deadline := time.Now().Add(closeDrain)
		_ = c.conn.SetReadDeadline(deadline)
		scratch := make([]byte, 4096)
		for time.Now().Before(deadline) {
			if _, err := c.conn.Read(scratch); err != nil {
				break
			}
		}
	}
	_ = c.conn.SetWriteDeadline(time.Now().Add(time.Second))
	return c.conn.Close()
}

File: reference/python/caret_v4/ws.py
"""RFC 6455, hand-rolled, standard library only.

Python ships no WebSocket implementation, and this repository takes no
third-party dependencies, so the framing lives here. It is the small
subset caret/v4 needs and no more: one connection, one message at a
time, no extensions, no compression, no subprotocols. Fragmented
messages are reassembled because a peer may send them; nothing here
sends one.

The same file serves both halves. The server side takes over a socket
that `http.server` has already read the request line from; the client
side dials, which is what the conformance checker uses to behave like a
keyboard.
"""
from __future__ import annotations

import base64
import hashlib
import os
import socket
import ssl
import struct
import time
from urllib.parse import urlsplit

WS_GUID = "258EAFA5-E914-47DA-95CA-C5AB0DC85B11"

OP_CONTINUATION = 0x0
OP_TEXT = 0x1
OP_BINARY = 0x2
OP_CLOSE = 0x8
OP_PING = 0x9
OP_PONG = 0xA

#: Refuse a frame larger than this outright. The protocol's own
#: max_frame_bytes is smaller and is enforced by the session; this is the
#: backstop that keeps a hostile length header from allocating a gigabyte.
MAX_PAYLOAD = 8 << 20

#: How long `close` spends reading the peer out before closing the
#: socket. A socket closed with unread inbound data is reset, and a reset
#: can discard the peer's receive buffer along with the terminal event
#: just written. §8 lets a backend answer a replayed operation from cache
#: while the client is still uploading, so this is an ordinary path.
CLOSE_DRAIN_SECONDS = 0.5


class WSError(Exception):
    """A malformed frame or a failed handshake."""


class CloseError(Exception):
    """The peer closed. `code` is the WebSocket close code."""

    def __init__(self, code: int, reason: str = "") -> None:
        super().__init__(f"websocket closed {code} {reason!r}")
        self.code = code
        self.reason = reason


def accept_key(key: str) -> str:
    """The Sec-WebSocket-Accept value for a client's key."""
    digest = hashlib.sha1((key + WS_GUID).encode("ascii")).digest()
    return base64.b64encode(digest).decode("ascii")


class Connection:
    """One WebSocket connection. Not safe for concurrent readers; writes
    are serialized by the socket's own `sendall`."""

    def __init__(self, sock: socket.socket, is_client: bool) -> None:
        self.sock = sock
        self.is_client = is_client
        self.reader = sock.makefile("rb")
        self._closed = False
        self._saw_close = False

    # ---------------------------------------------------------- reading

    def set_timeout(self, seconds: float | None) -> None:
        self.sock.settimeout(seconds)

    def _read_exactly(self, count: int) -> bytes:
        data = self.reader.read(count)
        if data is None or len(data) < count:
            raise WSError("connection closed mid-frame")
        return data

    def _read_frame(self) -> tuple[int, bytes, bool]:
        header = self._read_exactly(2)
        final = bool(header[0] & 0x80)
        if header[0] & 0x70:
            raise WSError("reserved bits set")
        opcode = header[0] & 0x0F
        masked = bool(header[1] & 0x80)
        length = header[1] & 0x7F
        if length == 126:
            length = struct.unpack("!H", self._read_exactly(2))[0]
        elif length == 127:
            length = struct.unpack("!Q", self._read_exactly(8))[0]
        if length > MAX_PAYLOAD:
            raise WSError(f"frame of {length} bytes exceeds the {MAX_PAYLOAD} byte ceiling")
        # RFC 6455 §5.1: a client masks, a server does not.
        if masked == self.is_client:
            raise WSError("frame masking does not match the connection role")
        mask = self._read_exactly(4) if masked else b""
        payload = self._read_exactly(length) if length else b""
        if masked and payload:
            payload = bytes(b ^ mask[i % 4] for i, b in enumerate(payload))
        if opcode in (OP_CLOSE, OP_PING, OP_PONG) and (not final or length > 125):
            raise WSError("control frames must be final and short")
        return opcode, payload, final

    def read_message(self) -> tuple[bool, bytes]:
        """The next data message as `(is_binary, payload)`.

        Control frames are handled here and never surface. A close frame
        raises `CloseError`, which is how every caller learns the
        operation is over.
        """
        assembled = bytearray()
        in_fragment = False
        binary = False
        while True:
            opcode, payload, final = self._read_frame()
            if opcode == OP_PING:
                self._write_frame(OP_PONG, payload)
                continue
            if opcode == OP_PONG:
                continue
            if opcode == OP_CLOSE:
                self._saw_close = True
                code = struct.unpack("!H", payload[:2])[0] if len(payload) >= 2 else 1005
                raise CloseError(code, payload[2:].decode("utf-8", "replace"))
            if opcode in (OP_TEXT, OP_BINARY):
                if in_fragment:
                    raise WSError("interleaved data frame")
                binary = opcode == OP_BINARY
                assembled = bytearray(payload)
            elif opcode == OP_CONTINUATION:
                if not in_fragment:
                    raise WSError("continuation without a start frame")
                assembled.extend(payload)
            else:
                raise WSError(f"unknown opcode {opcode:#x}")
            if final:
                return binary, bytes(assembled)
            in_fragment = True
            if len(assembled) > MAX_PAYLOAD:
                raise WSError("fragmented message exceeds the payload ceiling")

    # ---------------------------------------------------------- writing

    def _write_frame(self, opcode: int, payload: bytes) -> None:
        if self._closed:
            raise WSError("connection is closed")
        header = bytearray([0x80 | opcode])
        length = len(payload)
        mask_bit = 0x80 if self.is_client else 0x00
        if length < 126:
            header.append(mask_bit | length)
        elif length < (1 << 16):
            header.append(mask_bit | 126)
            header.extend(struct.pack("!H", length))
        else:
            header.append(mask_bit | 127)
            header.extend(struct.pack("!Q", length))
        if self.is_client:
            mask = os.urandom(4)
            header.extend(mask)
            payload = bytes(b ^ mask[i % 4] for i, b in enumerate(payload))
        self.sock.sendall(bytes(header) + payload)

    def send_text(self, text: str) -> None:
        self._write_frame(OP_TEXT, text.encode("utf-8"))

    def send_binary(self, data: bytes) -> None:
        self._write_frame(OP_BINARY, data)

    def send_ping(self, data: bytes = b"") -> None:
        self._write_frame(OP_PING, data)

    def close(self, code: int = 1000, reason: str = "") -> None:
        if self._closed:
            return
        self._closed = True
        payload = struct.pack("!H", code) + reason.encode("utf-8")[:123]
        try:
            self._closed = False  # _write_frame refuses a closed connection
            self._write_frame(OP_CLOSE, payload)
        except OSError:
            pass
        finally:
            self._closed = True
        if not self._saw_close:
            deadline = time.monotonic() + CLOSE_DRAIN_SECONDS
            try:
                while time.monotonic() < deadline:
                    self.sock.settimeout(max(0.01, deadline - time.monotonic()))
                    if not self.sock.recv(4096):
                        break
            except OSError:
                pass
        try:
            self.reader.close()
        except OSError:
            pass
        try:
            self.sock.close()
        except OSError:
            pass


# ------------------------------------------------------------ server side


def upgrade(handler) -> Connection:
    """Complete the handshake on a `BaseHTTPRequestHandler` and take the
    socket over. Raises `WSError` when the request is not a WebSocket
    upgrade, which the caller answers with a plain HTTP status."""
    headers = handler.headers
    if headers.get("Upgrade", "").lower() != "websocket":
        raise WSError("not a websocket upgrade")
    connection_tokens = [t.strip().lower() for t in headers.get("Connection", "").split(",")]
    if "upgrade" not in connection_tokens:
        raise WSError("missing Connection: Upgrade")
    if headers.get("Sec-WebSocket-Version", "") != "13":
        raise WSError("only WebSocket version 13 is supported")
    key = headers.get("Sec-WebSocket-Key", "")
    if not key:
        raise WSError("missing Sec-WebSocket-Key")

    response = (
        "HTTP/1.1 101 Switching Protocols\r\n"
        "Upgrade: websocket\r\n"
        "Connection: Upgrade\r\n"
        f"Sec-WebSocket-Accept: {accept_key(key)}\r\n\r\n"
    )
    handler.wfile.write(response.encode("ascii"))
    handler.wfile.flush()
    handler.close_connection = True
    return Connection(handler.connection, is_client=False)


# ------------------------------------------------------------ client side


def dial(url: str, headers: dict | None = None, timeout: float = 30.0,
         verify_tls: bool = True) -> tuple[Connection | None, int, str]:
    """Open a client connection.

    Returns `(connection, status, reason)`. On anything but 101 the
    connection is `None` and the status is the server's, so a caller can
    tell "refused my credential with 401" from "the socket died".
    """
    parts = urlsplit(url)
    secure = parts.scheme == "wss"
    port = parts.port or (443 if secure else 80)
    host = parts.hostname or "localhost"
    path = parts.path or "/"
    if parts.query:
        path += "?" + parts.query

    sock = socket.create_connection((host, port), timeout=timeout)
    if secure:
        context = ssl.create_default_context()
        if not verify_tls:
            context.check_hostname = False
            context.verify_mode = ssl.CERT_NONE
        sock = context.wrap_socket(sock, server_hostname=host)

    key = base64.b64encode(os.urandom(16)).decode("ascii")
    request = [
        f"GET {path} HTTP/1.1",
        f"Host: {host}:{port}" if parts.port else f"Host: {host}",
        "Upgrade: websocket",
        "Connection: Upgrade",
        f"Sec-WebSocket-Key: {key}",
        "Sec-WebSocket-Version: 13",
    ]
    for name, value in (headers or {}).items():
        request.append(f"{name}: {value}")
    sock.sendall(("\r\n".join(request) + "\r\n\r\n").encode("ascii"))

    reader = sock.makefile("rb")
    status_line = reader.readline().decode("latin-1").strip()
    parsed_headers = {}
    while True:
        line = reader.readline().decode("latin-1").strip()
        if not line:
            break
        name, _, value = line.partition(":")
        parsed_headers[name.strip().lower()] = value.strip()

    fields = status_line.split(" ", 2)
    status = int(fields[1]) if len(fields) > 1 and fields[1].isdigit() else 0
    reason = fields[2] if len(fields) > 2 else ""
    if status != 101:
        reader.close()
        sock.close()
        return None, status, reason
    if parsed_headers.get("sec-websocket-accept") != accept_key(key):
        reader.close()
        sock.close()
        raise WSError("the server's Sec-WebSocket-Accept does not match the key we sent")

    conn = Connection(sock, is_client=True)
    # Reuse the reader that already consumed the response headers, so no
    # bytes of the first frame are stranded in a discarded buffer.
    conn.reader = reader
    return conn, status, reason
